Response Quiz

JusticeV452 opened the pull request:
- Adds the ResponseQuiz view to the frontend, which enables playing the response quiz with a text in the database.
- In the backend, functions to create different types of errors are kept in 'error_functions' and are randomly applied to 3 answer choices for each question.
- Available errors include: capitalization error, comma splice error, replacement of a word with a homophone, run-on sentence error, verb conjugation error, and verb deletion error.
- The frontend is very similar to the verb_cojugation quiz and has a submit button that alerts the user about unanswered questions.
![image](https://user-images.githubusercontent.com/55002917/113627126-0c5e4280-9631-11eb-9c6c-b360f77217f2.png)
![image](https://user-images.githubusercontent.com/55002917/113627216-2ef05b80-9631-11eb-993d-12dde6339c2e.png)


- Upon pressing submit, each question will be updated to show the chosen answer, the correct answer, and a feedback text based on the different types of errors the chosen answer had. 
![image](https://user-images.githubusercontent.com/55002917/113627321-55ae9200-9631-11eb-9607-55ce3c06ac09.png)

- Currently, the quiz crashes if none of the errors can be applied to a certain question.

L-Jewel committed:
Copied the quiz module over and renamed

L-Jewel committed:
The page shows up now; will continue adapting it to fit the needs of our module soon

L-Jewel committed:
Adapted the module to handle multiple question types - as the user types into the response bar, their responses are instantly saved. Will add the grading part when the response validation is fully developed

JusticeV452 committed:
Added function to parse question and answer text

JusticeV452 committed:
Added verb conjugation error function and fixed get_quiz_questions

JusticeV452 committed:
Added a starting framework for applying errors to question answer choices

L-Jewel committed:
Merge branch 'master' into response-validation

L-Jewel committed:
Revert "Merge branch 'master' into response-validation"

This reverts commit 84501d5509fcaa54357030cf480ec0a01251785b, reversing
changes made to b17e8356d31e3a20c2b01c048fdbe81e014d11ab.

L-Jewel committed:
Revert "Revert "Merge branch 'master' into response-validation""

This reverts commit 96aaaaa21a599a03f25e13c4bd5aef790aa98c83.

L-Jewel committed:
updated the get_quiz_data view with the new db field name

L-Jewel committed:
Fixed the API urls for both of the quiz views

L-Jewel committed:
Added the endpoint that gets the quiz data

So far, it only displays the question--working on getting the answer choices to show up, as the data structure for the options has changed.

L-Jewel committed:
Fixed the bug with the MC display

JusticeV452 committed:
Changed 'apply_question_errors' to 'apply_question_option_errors' and updated functionality

Updated the api endpoint to add errors to the questions
Also fixed some warnings in views.py

jlin00 committed:
added comma splice error

jlin00 committed:
edited return statement

jlin00 committed:
updated nates

jlin00 committed:
updated fxn name

jlin00 committed:
updated comma splice error fxn

L-Jewel committed:
Fixed the scoring feature

L-Jewel committed:
Added a feature that provides the user with feedback based on the error types of the answer they choose.

L-Jewel committed:
Updated design to match the Figma

L-Jewel committed:
Fixed linter errors

L-Jewel committed:
Added a dictionary that maps error types to the appropriate error messages to be shown to the user

tcui37 committed:
created run_on error for conversation quiz

JusticeV452 committed:
add get_quiz_questions test

JusticeV452 committed:
Merge branch 'response-validation' of https://github.com/dhmit/lang_learn into response-validation

tcui37 committed:
Homophone Error

Added homophone error to conversation quiz file, which replaces a word with a homophone.

tcui37 committed:
Added similar-sounding-words to requirements

Added similar-sounding-words to requirements, containing a large dictionary of homophones for the homophones error for conversation quiz

tcui37 committed:
Updated feedback in Frontend

In the frontend file, added the run-on and homophone error descriptions as a part of machine generated feedback.

JusticeV452 committed:
Refactor conversation_quiz

- Moved all error functions to error function dir
in analysis folder
- changed question answer from index to the correct answer string (updated tests to follow these changes)

JusticeV452 committed:
Merge branch 'response-validation' of https://github.com/dhmit/lang_learn into response-validation

JusticeV452 committed:
Update conversation quiz and add new errors

- Prevent duplicate answer choices from being created
- Added capitalization error
- Added verb deletion error
- Changed run-on sentence error to work with '?' and '!'

JusticeV452 committed:
Merge branch 'master' into response-validation

JusticeV452 committed:
fixed linter error

JusticeV452 committed:
fixed missing trailing comma error

MBJean commented:
@JusticeV452 with PR #24 merged, you should be able to pull in the main branch and resolve the failing `python_linters` job here.

MBJean commented:
For the sake of coordinating timing, I'll have a full review fo this PR completed by noon tomorrow (4/8).

MBJean commented on backend/app/analysis/conversation_quiz.py:
```diff
@@ -0,0 +1,90 @@
+            if hasattr(error_function, 'apply'):
+                usable_functions.append(error_function)
+        except ModuleNotFoundError:
+            print(f'{error_name} error is not available.')
```
At first glance, this method of accessing the modules of the package `error_functions` seems strange to me, as it seems to require adding some overhead that is already built in to Python's package and module systems (splitting out the package name, rewriting the dot-syntax reference as a string, etc.). A more conventional pattern would be to import them explicitly via the package and add them to `useable_functions` by their imported name. What challenges led you to this implementation?

If a goal is to automatically include all future `error_functions` modules that have the `apply` function without needing to update this function, I'd argue that it's safer to require updating this function explicitly rather than letting it happen implicitly when a new module is created in `error_functions`. Otherwise, some future developer could create a new module in `error_functions` without necessarily realizing that it'll mutate some aspect of some other feature.

In any case, I'd love to hear more about what led you down this path.

MBJean commented on backend/app/analysis/conversation_quiz.py:
```diff
@@ -0,0 +1,90 @@
+            'options': [copy.deepcopy(answer)] * 4,
+            'answer': q_and_a[i + 1].strip(' '),
+        }
+        questions.append(new_question)
```
Looking ahead, I can see that we transform these newly created dictionaries using `apply_question_option_errors` in the API view endpoint. Did you consider calling `apply_question_option_errors` here in line 48? Doing so would allow us to do two things:

- Reduce the business logic of the API endpoint.
- More tightly couple the creation of this data structure (a dictionary with the keys `question`, `options`, and `answer`) with the function that will transform it. `apply_question_option_errors` in several places drills fairly deeply into the data structure created here in line 43 (`new_question`), so it'd aid in readability to call `apply_question_option_errors` immediately after the data structure it depends on.

MBJean commented on backend/app/views.py:
```diff
@@ -188,3 +189,16 @@ def get_quiz_data(request, text_id):
+    text_obj = Text.objects.get(id=text_id)
+    res = get_quiz_questions(text_obj.content)
+    for question in res:
+        apply_question_option_errors(question)
```
See my comment above about the possibility of moving this logic into `get_quiz_questions`.

MBJean commented on backend/app/analysis/error_functions/homophone.py:
```diff
@@ -0,0 +1,40 @@
+
+def apply(question_option):
+    """
+    Creates a new question option, but replacing one of the words with a homograph
```
Nitpick: `homograph` here should be `homophone`, right?

MBJean commented on backend/app/analysis/error_functions/run_on.py:
```diff
@@ -0,0 +1,40 @@
+        return question_option, False
+    sentences[selected_sentence_index] = new_sentence
+    # lowercase the next sentence
+    sentences[selected_sentence_index+1] = sentences[selected_sentence_index+1].lower()
```
This will lowercase the whole sentence, right? In `comma_splice` we use `sentence[0].lower() + sentence[1:]` to just lowercase the initial letter to avoid, say, lowercasing a proper name or something. Would something like that be worth introducing here?

MBJean commented on backend/app/tests.py:
```diff
@@ -160,3 +162,21 @@ def test_crossword(self):
+            ],
+            'answer': 'I am doing well. How are you?',
+        }]
+        self.assertEqual(questions, expected)
```
Thanks for writing a test around this part of the functionality! How challenging would it be to provide unit-level tests for each of the `error_functions` and one for `apply_question_option_errors`? I imagine that the unit tests would be relatively straightforward. The inherent randomness of `apply_question_option_errors` would present a challenge, but it'd probably be possible to test, at the very least, that the argument is mutated in some form (comparison check on the values in `quiz_question['options']`, possibly) as a result of calling it.

MBJean commented on frontend/src/responseQuizView/responseQuizView.css.map:
```diff
@@ -0,0 +1 @@
+{"version":3,"sourceRoot":"","sources":["quizView.scss"],"names":[],"mappings":"AAAQ;AAER;EACI;EACA;EACA;;;AAGJ;EACI;EACA;EACA;;;AAGJ;EACI;;;AAGJ;EACI;EACA;EACA","file":"quizView.css"}
```
Are these files (`responseQuizView.css` and `responseQuizView.css.map`) included intentionally? I'm surprised to see bare .css files and source maps getting checked in to production, and it looks like there's some duplication here with what's in `responseQuizView.scss`.

MBJean commented on frontend/src/responseQuizView/responseQuizView.scss:
```diff
@@ -0,0 +1,173 @@
+    border: none;
+    font-size: 28px;
+    font-weight: bold;
+}
```
In general, I'll recommend avoid selecting against HTML IDs for all the reasons well outlined [in this article](https://dev.to/clairecodes/reasons-not-to-use-ids-in-css-4ni4), but primarily because of the [inconsistency it introduces in CSS selector specificity](https://stuffandnonsense.co.uk/archives/css_specificity_wars.html). Front-end development is full of opinions, so you'll for instance find [this argument](https://stackoverflow.com/questions/1878810/is-there-any-pros-and-cons-if-i-use-always-css-class-instead-css-id-for-everythi) in favor of using IDs, so basically feel comfortable using whichever choice you all decide as team so long as you feel comfortable defending it.

In my experience, it's a useful frame of mind to get into to avoid mixing markup and style. For instances, avoid giving elements base styling:
```
h1 {
  ...
}
````

But rather supply a styled class:
```
.h1 {
  ...
}
```

This bulks up the HTML with more classes, but HTML primarily serves as a structuring and accessibility tool, not a design tool. In the future, a designer might request applying `h1` styling to a non `h1` element, so it's useful to have already been in the frame that all styling is class based. I lump selecting against HTML IDs in the same category: HTML IDs are anchors in the DOM, selecting via `<a>` tags and useful for having a reference point in a URL (`www.example.com/about#somewhere-in-the-page`), so I've found it useful to separate out styling from their functionality.

MBJean commented on frontend/src/responseQuizView/responseQuizView.js:
```diff
@@ -0,0 +1,343 @@
+    'comma-splice': 'Two independent clauses are incorrectly joined by a comma',
+    'run-on': 'There is a run on sentence',
+    'homophone': 'Misuse of a similar sounding word (homophone)',
+};
```
I'd recommend not using PascalCase for the naming of this object. PascalCase is generally reserved in JavaScript for components, classes, and interfaces (as [this article ](https://www.robinwieruch.de/javascript-naming-conventions)outlines). This is by no means a hard and fast rule, just a convention among front-end developers that's worth following if you engage with the broader development community.

L-Jewel replied:
Renamed to "ERROR_DESCRIPTIONS" to prevent from using PascalCase and because the dictionary being created here is a constant.

MBJean commented on frontend/src/responseQuizView/responseQuizView.js:
```diff
@@ -0,0 +1,343 @@
+        try {
+            const response = await fetch(`/api/get_response_quiz_data/${this.props.textID}/`);
+            const data = await response.json();
+            this.setState({ data: data });
```
This is a relatively minor point, but I'd recommend naming this something more descriptive than `data`. Just reading this `componentDidMount` offers no information about the shape of this data.

L-Jewel replied:
Renamed to "questionData" for the sake of clarity.

MBJean commented on frontend/src/responseQuizView/responseQuizView.js:
```diff
@@ -0,0 +1,343 @@
+                question: prevQuestionNumber,
+            });
+        } else {
+            console.log();
```
It looks like this and the above (`nextQuestion`) might be unfinished, or at the very least could have their `else` blocks removed.

L-Jewel replied:
The else statements have been removed

MBJean commented on frontend/src/responseQuizView/responseQuizView.js:
```diff
@@ -0,0 +1,343 @@
+
+    onProgressBarClick = (event) => {
+        this.setState({
+            ...this.state,
```
What's the purpose of spreading `this.state` here?

L-Jewel replied:
Removed.

MBJean commented on frontend/src/responseQuizView/responseQuizView.js:
```diff
@@ -0,0 +1,343 @@
+
+    onAnswerChoiceClick = (event) => {
+        const currentQ = this.state.question;
+        if ((typeof event.target.value) === 'undefined') {
```
In which circumstances would the `value` here be `undefined`?

L-Jewel replied:
If I remember correctly, for some reason, when answer choices are clicked, it would return an undefined event right next to an event with an actual value. I can check and see if this is still the case, though.

L-Jewel replied:
This is still the case. I'm not entirely sure why, but an undefined event happens immediately before a substantive one. The current implementation is to make sure nothing breaks when this undefined event is caught.

However, the if-statement has since been simplified into one (rather than if...else if).

MBJean commented on frontend/src/responseQuizView/responseQuizView.js:
```diff
@@ -0,0 +1,343 @@
+        } else if (!this.state.graded) {
+            const answers = this.state.userAnswers;
+            answers[currentQ] = event.target.value;
+            this.setState({ userAnswers: answers });
```
I think this could use a bit of a refactor. It looks like this is mutating state twice: since `this.state.userAnswers`  is not a primitive type, `const answers = this.state.userAnswers` doesn't make a copy but rather simply assigns the reference. `answers[currentQ] = event.target.value` therefore updates `this.state.userAnswers` without triggering a rerender (which occurs when you call `this.setState`). The conventional pattern to copy out the object before setting it would look something like:
```
this.setState({
  userAnswers: {
    ...this.state.userAnswers,
    [currentQ]: event.target.value
  }
})
```

MBJean reviewed (changes requested):
Great work, everyone! There's a huge amount of complexity to this new feature and you've handled it with fewer lines of code than I would have expected (especially on the back end).

There are two places I still want to spend more review on: the contents of the new React component's `render` method, and the CSS. That said, I've included the bulk of my comments below. I'll have the remainder of my comments available to you tomorrow. I've not identified significant blockers other than one React implementation that probably needs a refactor and at least the consideration of a few unit-style tests on the modules in `error_functions`.

With all of my reviews, please always feel comfortable pushing back or explaining your reasoning. If anyone has any questions or needs clarification, don't hesitate to reach out and we can meet in office hours, the Friday hack session, or schedule some specific time to meet.

MBJean commented on frontend/src/responseQuizView/responseQuizView.scss:
```diff
@@ -0,0 +1,173 @@
+::-webkit-scrollbar-thumb {
+    background: #E5D4ED;
+    border-radius: 10px;
+}
```
This is by no means a blocker, but if you're interested in writing future-proof and browser-compatible CSS, here's a few articles:

[Using relative units for sizing](https://www.24a11y.com/2019/pixels-vs-relative-units-in-css-why-its-still-a-big-deal/). This is a big aid in responsive design and worth thinking about, if you haven't already.

[Alphabetical ordering for CSS properties](https://medium.com/@jerrylowm/alphabetize-your-css-properties-for-crying-out-loud-780eb1852153). This is a big aid in avoiding headaches when you inevitably need to update old CSS.

MBJean commented on frontend/src/responseQuizView/responseQuizView.js:
```diff
@@ -0,0 +1,343 @@
+                                                }
+                                            }
+                                        }
+                                    }
```
For the conditionals on lines 220 and 234, what led you to using `Object.prototype.hasOwnProperty.call`? As far as I can tell, `this.state.userAnswers` shouldn't have a `null` prototype, which is the use case for this elaborate set of invocations outlined in [this article](https://dev.to/aman_singh/what-s-the-deal-with-object-prototype-hasownproperty-call-4mbj). 

The more straightforward alternatives might be: `answers.hasOwnProperty(someValue)` or `someValue in answers`. While the latter of these [does check inherited properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in), in both of these cases we're looking for an integer as the object key and so we should be reasonably safe to use any of those options (i.e., `1` isn't going to be an inherited property of the Object instance `this.state.userAnswers`).

If you have a second, I'd be interested in hearing more about what challenges led you down that path.

L-Jewel replied:
Initially, something far out of my control wouldn't let me run the site without wrapping that code block in that if-statement. However, I've recently removed it, and it appears that things work fine now without it. It's been replaced with `if (qNumber in answers)`

MBJean commented on frontend/src/responseQuizView/responseQuizView.js:
```diff
@@ -0,0 +1,343 @@
+                            }
+                        </div>
+                        <div className="col-9 shaded-box">
+                            <h3 className="question-title">Question #{this.state.question}</h3>
```
Unless I missed it, I think this `<h3>` should be an `<h2>` to follow the correct order of header hierarchy (i.e., `<h1>` should be followed by an `<h2>`, etc.).

MBJean commented on frontend/src/responseQuizView/responseQuizView.js:
```diff
@@ -0,0 +1,343 @@
+                                            }
+                                        </p>
+                                    </p>
+                                    <b className="results-header">Machine-Generated Feedback</b>
```
This `<b>` tag looks like it could more semantically be rendered as a header tag of some kind. The [usage notes for the `<b>` tag](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/b#usage_notes) warn against impulse to use `<b>` in place of header tag.

MBJean reviewed (commented):
Here's one final pass of review feedback. Overall the React `render` code looks good, as does the CSS. I leave a few remarks here that I'd be interested in your feedback on.

L-Jewel committed:
Removed the CSS and CSS map files

L-Jewel committed:
renamed a few of the variables

"Data" wasn't a very descriptive, so we renamed it to "questionData".

"ErrorDescriptions" was renamed to "ERROR_DESCRIPTIONS" to prevent from using PascalCase

L-Jewel committed:
Error handling for invalid texts

Gives the user an error page whenever they try to access the quiz or response quiz module with an invalid text

L-Jewel committed:
Added an error description for verb deletion

L-Jewel committed:
Removed unnecessary else statements from next/prev question functions

L-Jewel committed:
Replaced the Object.prototype if-statements with more straightforward arguments

L-Jewel committed:
Removed the spreading of this.state in onProgressBarClick()

L-Jewel committed:
Consolidated the if statement within onAnswerChoiceClick()

L-Jewel committed:
Resolved redundant code within onAnswerChoiceClick

L-Jewel committed:
Addressed header tag order semantics

JusticeV452 committed:
Merge branch 'master' into response-validation

JusticeV452 committed:
Made requested changes

- Changed dynamic error function imports to explicit imports
- Moved application of errors to get_quiz_question function
- Added unit tests for apply_question_option_errors and all the error_functions
- Removed test for get_quiz_questions

JusticeV452 committed:
fixed too-many-blank-lines

JusticeV452 committed:
Fixed linter errors

- Added docstrings to test cases and error function modules

JusticeV452 committed:
Added docstrings

MBJean reviewed (approved):
Thanks everyone for addressing my feedback! This LGTM, and I'll leave my approval now while you address the final few linting issues.

MBJean committed:
Add eslint disable for alert popup
